src/Authoring/WinRT.SourceGenerator/WinRTTypeWriter.cs

```diff
@@ -1,2180 +1,2682 @@
-                }
-
-                AddCustomAttributes(node.GetAttributes(), parentHandle);
-            }
-        }
-
-        public bool IsPublicNode(MemberDeclarationSyntax node)
-        {
-            return node.Modifiers.Any(modifier => modifier.ValueText == "public") ||
-                (node.Parent is InterfaceDeclarationSyntax && node.Modifiers.Count == 0);
-        }
-
-        public bool IsPrivateNode(AccessorDeclarationSyntax node)
-        {
-            return node.Modifiers.Any(modifier => modifier.ValueText == "private");
-        }
-
-        public bool IsStaticNode(MemberDeclarationSyntax node)
-        {
-            return node.Modifiers.Any(modifier => modifier.ValueText == "static");
-        }
-
-        public bool IsSealedNode(MemberDeclarationSyntax node)
-        {
-            return node.Modifiers.Any(modifier => modifier.ValueText == "sealed");
-        }
-
-        public string GetNamespace()
-        {
-            return string.Join(".", namespaces);
-        }
-
-        public string QualifiedName(string identifier)
-        {
-            return QualifiedName(GetNamespace(), identifier);
-        }
-
-        public string QualifiedName(string @namespace, string identifier)
-        {
-            return string.Join(".", @namespace, identifier);
-        }
-
-        public override void VisitNamespaceDeclaration(NamespaceDeclarationSyntax node)
-        {
-            namespaces.Push(node.Name.ToString());
-            base.VisitNamespaceDeclaration(node);
-            namespaces.Pop();
-        }
-    }
-}
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Generator
+{
+    class Parameter
+    {
+        public Symbol Type;
+        public string Name;
+        public ParameterAttributes Attributes;
+        public bool ByRef;
+
+        public Parameter(Symbol type, string name, ParameterAttributes attributes)
+            :this(type, name, attributes, attributes == ParameterAttributes.Out)
+        {
+        }
+
+        public Parameter(Symbol type, string name, ParameterAttributes attributes, bool byRef)
+        {
+            Type = type;
+            Name = name;
+            Attributes = attributes;
+            ByRef = byRef;
+        }
+
+        public Parameter(ITypeSymbol type, string name, ParameterAttributes attributes)
+            : this(new Symbol(type), name, attributes)
+        {
+        }
+
+        public Parameter(EntityHandle type, string name, ParameterAttributes attributes)
+            : this(new Symbol(type), name, attributes)
+        {
+        }
+
+        public Parameter(IParameterSymbol parameterSymbol)

```

Comment:
could you add some comments? I don't quite see what this function is doing. I think it is doing the check on valid read/write only array attribute usage?

Reply:
This is just a constructor for the scenarios where we have IParameterSymbol as what we use to construct this object.  The attribute check is there to make sure we consult it to determine whether to make the param out or in.  But I will add a comment.